aggregator/handler/router/with_ack.go

```diff
@@ -0,0 +1,48 @@
+package router
+
+import (
+	"github.com/m3db/m3aggregator/aggregator/handler/common"
+	"github.com/m3db/m3msg/producer"
+)
+
+type withAckRouter struct {
+	p producer.Producer
+}
+
+// NewWithAckRouter creates a new router that routes buffer and waits for acknowledgements.
+func NewWithAckRouter(p producer.Producer) Router {
+	return withAckRouter{p: p}
+}
+
+func (r withAckRouter) Route(shard uint32, buffer *common.RefCountedBuffer) error {
+	return r.p.Produce(newMessage(shard, buffer))
+}
+
+func (r withAckRouter) Close() {
+	r.p.Close(producer.WaitForConsumption)
+}
+
+type message struct {
+	shard  uint32
+	buffer *common.RefCountedBuffer
+}
+
+func newMessage(shard uint32, buffer *common.RefCountedBuffer) producer.Message {
```

Comment:
This leads to one allocation per call due to casting to the interface, do you need to pool this?

Reply:
The message pool in producer creates 3X(3 consumers) objects whenever one of this message here was created, and the result shows using allocation there is more efficient than pooling, so I feel we probably don't need to pool this.

Reply:
I think pooling is still useful when the number of objects needed / s is not too high and there aren't a large number of goroutines trying to get them at the same time. Empirically speaking the agg tier produces roughly 8M metrics / s, so roughly 800K buffers / s, given the agg tier size, that's roughly 4K buffers / s per instance, IOW you'll have about 16 goroutines contending for a total of 4K buffers / s during flushing, which is not a lot of contention which is why I think pooling might actually help here.

You might be right that that's still too much contention overhead for the agg tier, but I'd like to see some empirical evidence with production agg tier load as opposed to microbenchmarks. If you intend to measure this in the future, can you either create an issue and/or add a TODO here?

Reply:
Sure, can add a pool here if needed, will add a TODO for now.